### DISCUSSION TITLE
Cross compiling to aarch64-linux-android

### DISCUSSION THREAD

**Discussion by rem821** [eclipse-zenoh/roadmap#98] [Q&A]
Hi,

I've been wondering, whether it would be possible to cross-compile to aarch64-linux-android. I'm currently building a VR app in C++ that should be able to listen on some ROS2 topics and visualize them. The readme discusses cross-compiling to aarch64-unknown-linux-gnu which is working fine, but when I try to do an Android build, I'm getting errors in the linker stage.

```
  /home/standa/Android/Sdk/ndk/21.4.7075529/toolchains/llvm/prebuilt/linux-x86_64/lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: _deps/zenohc-build/debug/target/debug/libzenohcd.a(zenohcd.2r27w0emrex4v6rh.rcgu.o): Relocations in generic ELF (EM: 62)
  _deps/zenohc-build/debug/target/debug/libzenohcd.a: error adding symbols: File in wrong format
```

Sorry, I'm not exactly familar with how Rust libraries compile in cmake, so I'm a bit confused.

**Comment by DariusIMP** [comment]
Hi, sorry for the late reply. I understand you need to build zenoh-c for Android?
We could investigate that, it should be possible, although it's a bit tricky to build Zenoh-C for Android, since the Rust components need to be built using the Android NDK. 
Out of curiosity, what framework are you using for your VR app?

Btw, we recently released an alpha version of a Kotlin bindings for Zenoh that work very well for Android (https://github.com/eclipse-zenoh/zenoh-kotlin). I don't know the internals of your project but maybe you want to take a look into it, hopefully it can be helpful. There, to build the native libraries, _we*_ use the Rust Android Gradle plugin which uses the Android NDK.

----
_*edit note: **we** use the Rust Android Gradle plugin..._

**Comment by rem821** [comment]
Hi, no worries, thanks for taking your time. 
Yes, as a part of my research project, I'm trying to hook onto ROS2 communication from inside Oculus Quest 2 headset. Basically in whatever way excluding having a DDS-UDP bridge that I would have to modify each time a topic is added.

The app is written in C++ using OpenXR to interface with the device, so no Kotlin sadly (but maybe a JNI bridge from Kotlin to C++ would be possible but I smell a lot of work coming from this side).

What I really like about zenoh compared to DDS implementations used in ROS2 is the ability to route outside of LAN, but this is not a strong requirement from the project. 

So far I've tried cross-compiling to aarch64-linux-android even though readme says it is not supported, but aarch64-unknown-linux-gnu is and I figured it maybe could be done though I'm currently not knowledgeable enough.

Another thing I've considered is to try to use zenoh-pico, which is natively C, but this implementation also doesnt support Android builds at the moment.

**Reply by JEnoch** [reply]
Hi,

> I'm trying to hook onto ROS2 communication from inside Oculus Quest 2 headset. Basically in whatever way excluding having a DDS-UDP bridge that I would have to modify each time a topic is added.

Did you have a look to our [Zenoh/ROS 2 bridge ](https://github.com/eclipse-zenoh/zenoh-plugin-ros2dds)?
It automatically discovers all the ROS pub/sub/services/actions at runtime and establishes routes to Zenoh pub/sub/queryables. No need to modify anything (unless you want to tune in its config file the allowed/denies topics or the republication rate for downsampling).

Note that Zenoh also provides a REST API (activatable in any Zenoh router or in the ROS 2 bridge). Even if not wire-efficient, you can publish/subscribe/query via HTTP up to ROS 2 with this. That could be a temporary alternative, waiting for a cross-build solution.

**Reply by rem821** [reply]
Hi, thanks for the suggestions! REST API surely looks like a good temporary solution.
The problem is, and that's why I'm trying to go the Zenoh way on the Oculus Quest 2 headset, that I don't currently have a working solution to cross-compile the whole ROS2 stack into android. Though I'm curious about the Zenoh/ROS 2 bridge suggestion - Is that something I could just link to my app running on Oculus and be able to discover ROS2 topics? Looking through the release build artifacts, the aarch64-linux-android build isn't provided. 

I've also stumbled upon [this](https://github.com/sloretz/sensors_for_ros/tree/master) project that managed to cross-compile the whole ROS2 with cyclone-dds to Android, so it looks like my next best shot

**Reply by JEnoch** [reply]
Sorry, I think I misunderstood your use case. My understanding was you have a ROS 2 system (robot?) running on some non-Android hosts, and you would like an Android-based Oculus Quest 2 subscribing to ROS 2 publications.
For this use case, you wouldn't necessary need a ROS 2 stack in Oculus Quest 2. You could just deploy a `zenoh-bridge-ros2dds` in the robot, and in Oculus Quest 2 have a pure-Zenoh application subscribing to the topics routed by the bridge.
Moreover, if as I 'm guessing the communication with Oculus Quest 2 goes over WiFi, this would avoid you all the troubles lot of ROS 2 users are experimenting with [ROS2/DDS discovery traffic over WiFi](https://zenoh.io/blog/2021-03-23-discovery/)

Now, if you need the full ROS 2 stack to run on Android in Oculus Quest 2, that's another story. 
Android is not listed as a [supported platform for ROS 2](https://ros.org/reps/rep-2000.html), and I'm not aware of any attempt for this, except the project you just introduced to me.

> Though I'm curious about the Zenoh/ROS 2 bridge suggestion - Is that something I could just link to my app running on Oculus and be able to discover ROS2 topics? 

The bridge is written in Rust, building and statically linking with CycloneDDS (no easy dynamic linking possibility with Rust)
It's not targeting Android, as we were not thinking of any use case of ROS2/DDS running on Android. Most Android devices are relying on wireless connectivity, so we don't recommend DDS usage on it, except for limited scale of DDS Readers/Writers and finely tuned DDS configuration and QoS.

On the other hand, if you can afford to wait for [ROS 2 Jazzy](https://docs.ros.org/en/foxy/Releases/Release-Jazzy-Jalisco.html) (May 2024), the development of [rmw_zenoh](https://github.com/ros2/rmw_zenoh) just started. That would eventually allow direct Zenoh communication between ROS 2 Nodes in your Robot and in Oculus Quest 2. 😃

**Reply by rem821** [reply]
> My understanding was you have a ROS 2 system (robot?) running on some non-Android hosts, and you would like an Android-based Oculus Quest 2 subscribing to ROS 2 publications.

That's exactly my use case, though, before discovering Zenoh, I've made several attempts to compile the whole ROS2 into the headset. Running a pure-Zenoh app on the headset is equivalently good solution - but that brings me to the original question. How to compile zenoh-c (preferably with zenoh-cpp wrapper), or zenoh-pico onto aarch64-linux-android architecture? Whichever is the easiest.

> On the other hand, if you can afford to wait for [ROS 2 Jazzy](https://docs.ros.org/en/foxy/Releases/Release-Jazzy-Jalisco.html) (May 2024), the development of [rmw_zenoh](https://github.com/ros2/rmw_zenoh) just started

Glad to hear that, from what I've seen online, this is going to allow for many new uses cases and I'm really fan of that.

**Reply by JEnoch** [reply]
> How to compile zenoh-c (preferably with zenoh-cpp wrapper), or zenoh-pico onto aarch64-linux-android architecture?

Indeed. We don't know yet how to do this, but we'll try to figure out and keep you posted.
